R/modules.R

```diff
@@ -121,23 +121,40 @@ createSessionProxy <- function(parentSession, ...) {
 #'
 #' @export
 moduleServer <- function(id, module, session = getDefaultReactiveDomain()) {
+  if (inherits(session, c("MockShinySession"))) session$isModuleServer <- TRUE
   callModule(module, id, session = session)
 }
 
-
 #' @rdname moduleServer
 #' @export
 callModule <- function(module, id, ..., session = getDefaultReactiveDomain()) {
-  if (!inherits(session, "ShinySession") && !inherits(session, "session_proxy")) {
-    stop("session must be a ShinySession or session_proxy object.")
+  if (!inherits(session, c("ShinySession", "MockShinySession", "session_proxy"))) {
+    stop("session must be a ShinySession, MockShinySession, or session_proxy object.")
   }
-  childScope <- session$makeScope(id)
-
-  withReactiveDomain(childScope, {
-    if (!is.function(module)) {
-      stop("module argument must be a function")
-    }
 
-    module(childScope$input, childScope$output, childScope, ...)
-  })
+  if (inherits(session, "MockShinySession") && session$isModuleServer) {
+    # If the module is under test *and* was called by moduleServer(), modify the
+    # module function locally by inserting the equivalent of `session$env <-
+    # environment()` at the beginning of its body. A similar operation is
+    # performed by .testModule() if the module is *not* called through
+    # moduleServer() but is under test. See .testModule() for details.
+    body(module) <- rlang::expr({
+      base::assign("env", base::environment(), envir = !!session)
+      !!!body(module)
+    })
+    isolate(
```

Comment:
Is this `isolate(withReactiveDomain(..., with_options` stuff necessary? That stuff is also present in test-module.R, line 106.